src/Npgsql/NpgsqlParameter.cs

```diff
@@ -368,7 +368,14 @@ public NpgsqlDbType NpgsqlDbType
             {
                 if (_dataTypeName != null)
                     return _dataTypeName;
-                throw new NotImplementedException("Infer from others");
+                if (_npgsqlDbType.HasValue)
+                {
+                    var primitiveNpgsqlType = _npgsqlDbType.Value & ~NpgsqlDbType.Array & ~NpgsqlDbType.Range;
```

Comment:
Returning the element type name when the NpgsqlDbType is an array/range does not seem like a good idea to me... `text` and `text[]` are two completely different types, so we should not be returning `text` for `NpgsqlDbType.Text | NpgsqlDbType.Array`

Reply:
Good point! Should we return `text[]` for `NpgsqlDbType.Text | NpgsqlDbType.Array`? What about `NpgsqlDbType.Text | NpgsqlDbType.Range` should it also be `text[]` or is there another representation for range?

Reply:
> Should we return text[] for NpgsqlDbType.Text | NpgsqlDbType.Array?

Yes.

> What about NpgsqlDbType.Text | NpgsqlDbType.Range should it also be text[] or is there another representation for range?

No - ranges and arrays are completely distinct types, so that would be an incorrect inference.

Unlike arrays, range types have to be manually created over their subtypes (aside from a handful of built-in range types), so they're similar to enums and composite types in that respect. In other words, for every PG base type `x`, there exists `x[]` for an array, but there isn't a similar thing for ranges.

This means that we can't know - at the global type mapper level - which PG type corresponds to `NpgsqlDbType.Text | NpgsqlDbType.Range`, and so we should fail to infer and return null (we should test that behavior).